src/utils/futures.rs

```diff
@@ -280,6 +281,39 @@ pub async fn delay(duration: Duration) {
         .ok();
 }
 
+/// Time a future execution and expose using metrics.
+///
+/// The future is expected to resolve to a Result and the status is logged as well.
+pub async fn time_task<F, T, E>(task_name: &str, future: F) -> F::Output
+where
+    F: Future<Output = Result<T, E>>,
+{
+    let start_time = Instant::now();
+    let ret = future.await;
+    metric!(
+        timer("futures.done") = start_time.elapsed(),
+        "task_name" => task_name,
+        "status" => if ret.is_ok() { "ok" } else {"err"},
+    );
+    ret
+}
+
+/// Retry a future 3 times with exponential backoff.
+pub async fn retry<G, F, T, E>(task_gen: G) -> Result<T, E>
```

Comment:
_nit_: Can you add unit tests for this function? There are great facilities that allow to mock the timer.

Reply:
not going to happen for now

Reply:
👍  Agreed, this is easier with newer tokio.